### PR TITLE
refactor(navigation): refactor navigation to be able to navigate in a separate window

### PR DESCRIPTION
**Pull request by calahatras**
service provider scope is changed so that a scope is created per window instead of per control, and each NavigationService creates its own Window instance

### CODE REVIEW

File: GitOut/Features/Navigation/NavigationService.cs [hunk 1]
```diff
@@ -18,41 +19,36 @@ public class NavigationService : INavigationService
         private readonly IServiceProvider provider;
         private readonly ITitleService titleService;
         private readonly IThemeService theme;
-        private readonly IOptions<NavigationWindowOptions> windowOptions;
-        private readonly IWritableStorage storage;
         private readonly ILogger<NavigationService> logger;
-        private readonly Stack<Tuple<ContentControl, string?, IServiceScope>> pageStack = new();
+
+        private readonly Stack<Tuple<ContentControl, string?>> pageStack = new();
```
Comment (mickedefault) [inline GitOut/Features/Navigation/NavigationService.cs]: can use a named tuple instead `(ContentControl current, string? title)`

File: GitOut/Features/Navigation/NavigationService.cs [hunk 2]
```diff
@@ -74,12 +72,12 @@ public void Back()
...
             }
-            (ContentControl current, string _, IServiceScope scope) = pageStack.Pop();
-            (ContentControl latest, string? title, IServiceScope _) = pageStack.Peek();
+            (ContentControl current, string _) = pageStack.Pop();
+            (ContentControl latest, string? title) = pageStack.Peek();
             OnNavigationRequested(latest);
             titleService.Title = title;
             CurrentPage = latest.GetType().FullName;
-            logger.LogInformation(LogEventId.Navigation, $"Navigating back to {CurrentPage} ({title})");
+            logger.LogInformation(LogEventId.Navigation, "Navigating back to {CurrentPage} ({Title})", CurrentPage, title);
```
Comment (mickedefault) [inline GitOut/Features/Navigation/NavigationService.cs]:
```suggestion
            logger.LogInformation(LogEventId.Navigation, $"Navigating back to {CurrentPage} ({Title})", CurrentPage, title);
```

**Review by mickedefault — COMMENTED** [review]
if you ctrl click settings and close the original, there's no way to get back :smile: similarly with a stage view
maybe the new window can inherit the current history from the spawning window?
ctrl clicking stageview and closing the new window triggers an exception for [watcher.EnableRaisingEvents](https://github.com/calahatras/gitout/blob/main/GitOut/Features/IO/GitRepositoryFileSystemWatcherProvider.cs#L35)

File: GitOut/Features/Navigation/NavigationService.cs (hunk 2 above)
Comment (calahatras) [reply GitOut/Features/Navigation/NavigationService.cs]: Apparently not:
https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2254
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-6.0#log-level
As I understand it, the reason is two-fold, the logger formats the string anyway by using the attached properties, and by using a constant log string and passing the values as items, the logger can use the names in the log format string to name the variables passed to it, in Application Insights these are then bound to a property bag called `customDimensions` that you can query specifically.

File: GitOut/Features/Navigation/INavigationService.cs [hunk 3]
```diff
@@ -1,13 +1,17 @@
...
     {
         string? CurrentPage { get; }
+
+        event EventHandler<EventArgs>? Closed;
         event EventHandler<NavigationEventArgs> NavigationRequested;
+
         T? GetOptions<T>(string pageName) where T : class;
 
-        void Navigate(string page, object? options);
+        void Navigate(string page, object? options, NavigationOptions? navigation = default);
```
Comment (mickedefault) [inline GitOut/Features/Navigation/INavigationService.cs]: do we need the default parameter? it's only used at one call site

File: GitOut/Features/Navigation/NavigationService.cs [hunk 4]
```diff
@@ -88,112 +80,124 @@ public void Back()
...
+            {
+                return false;
+            }
 
-        public void Navigate(string pageName, object? options)
+            ContentControl control = pageStack.Peek().Control;
+            return Window.GetWindow(control) == Application.Current.MainWindow && (pageStack.Count > 1 || control.DataContext is INavigationFallback);
+        }
+
+        public void Navigate(string pageName, object? options, NavigationOptions? navigation = default)
```
Comment (mickedefault) [inline GitOut/Features/Navigation/NavigationService.cs]: same here with the default parameter

**Review by mickedefault — COMMENTED** [review]
tested and works like a charm :star:

**Review by mickedefault — APPROVED** [review]
⭐

**Pull request merged**